File: golang/search_queue.go
package main

import (
	"fmt"
	"strings"
)

var graph = map[string][]string{
	"you":    {"alice", "bob", "claire"},
	"bob":    {"anuj", "peggy"},
	"alice":  {"peggy"},
	"claire": {"thom", "jonny"},
	"anuj":   {},
	"peggy":  {},
	"thom":   {},
	"jonny":  {},
}

func search(name string) bool {
	search_queue := []string{}
	search_queue = append(search_queue, graph[name]...)
	searched := make(map[string]bool, 0)

	for len(search_queue) != 0 {
		var person string
		person, search_queue = search_queue[0], search_queue[1:]

		if searched[person] {
			continue
		}

		if personIsSeller(person) {
			fmt.Println(person + " is a mango seller!")
			return true
		} else {
			search_queue = append(search_queue, graph[person]...)
			searched[person] = true
		}
	}

	return false
}

func personIsSeller(name string) bool {
	return strings.HasSuffix(name, "m")
}

func callSearch() {
	search("you")
}

File: python/search_queue.py
from collections import deque


graph = {}
graph["you"] = ["alice", "bob", "claire"]
graph["bob"] = ["anuj", "peggy"]
graph["alice"] = ["peggy"]
graph["claire"] = ["thom", "jonny"]
graph["anuj"] = []
graph["peggy"] = []
graph["thom"] = []
graph["jonny"] = []


def search(name):
    search_queue = deque()
    search_queue += graph[name]
    searched = {}  # Этот используется для отслеживания уже проверенных людей

    while search_queue:
        person = search_queue.popleft()

        if searched.get(person, None):
            continue

        if person_is_seller(person):
            print(person + " is a mango seller!")
            return True
        else:
            search_queue += graph[person]
            searched[person] = True
    return False


def person_is_seller(name):
    return name[-1] == "m"


if __name__ == "__main__":
    search("you")
